File: 102.cpp
// https://leetcode.com/problems/binary-tree-level-order-traversal/
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> levelOrder(TreeNode* root) {
        if (!root) return {};
        std::vector<std::vector<int>> result;
        std::queue<TreeNode*> nodeQueue;
        nodeQueue.push(root);
        nodeQueue.push(nullptr);
        
        std::vector<int> levelData;
        while (!nodeQueue.empty()) {
            TreeNode* ptr = nodeQueue.front();
            nodeQueue.pop();
            
            // level ends
            if (ptr == nullptr) {
                result.push_back(std::move(levelData));
                continue;
            }
            levelData.push_back(ptr->val);
            bool addSon = false;
            if (ptr->left != nullptr) {
                nodeQueue.push(ptr->left);
            }
            if (ptr->right != nullptr) {
                nodeQueue.push(ptr->right);
            }
            if (nodeQueue.size() != 1 && nodeQueue.front() == nullptr) nodeQueue.push(nullptr);
        }
        return result;
    }
};

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> levelOrder(TreeNode* root) {
        if (!root) return {};
        std::vector<std::vector<int>> result;
        std::queue<TreeNode*> nodeQueue;
        nodeQueue.push(root);
        
        int numsOfLevel = 1;
        std::vector<int> levelData;
        while (!nodeQueue.empty()) {
            for (int i = 0; i < numsOfLevel; ++i) {
                TreeNode* ptr = nodeQueue.front();
                nodeQueue.pop();
                levelData.push_back(ptr->val);
                if (ptr->left != nullptr) nodeQueue.push(ptr->left);
                if (ptr->right != nullptr) nodeQueue.push(ptr->right);
            }
            result.push_back(std::move(levelData));
            numsOfLevel = nodeQueue.size();
        }
        return result;
    }
};

File: golang/102.go
package main

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

func levelOrder(root *TreeNode) (ans [][]int) {
	if root == nil {
		return
	}

	queue := []*TreeNode{}
	queue = append(queue, root)
	queue = append(queue, nil)
	leveldata := []int{}

	for len(queue) != 0 {
		front := queue[0]
		queue = queue[1:]

		if front == nil {
			ans = append(ans, leveldata)
			leveldata = []int{}

			continue
		}

		leveldata = append(leveldata, front.Val)
		if front.Left != nil {
			queue = append(queue, front.Left)
		}
		if front.Right != nil {
			queue = append(queue, front.Right)
		}
		if len(queue) != 1 && queue[0] == nil {
			queue = append(queue, nil)
		}
	}
	return
}
